### PR TITLE
create deploy-to-aks action

### PR DESCRIPTION
**Pull request by RMcVelia**
## Context
Create AKS deploy action

## Changes proposed in this pull request
Add deploy-to-aks action and Readme.
Add slack failure notification to build action

## Guidance to review
see https://github.com/DFE-Digital/get-a-teacher-relocation-payment/actions/runs/11616882883 run from https://github.com/DFE-Digital/get-a-teacher-relocation-payment/pull/447

## Checklist

- [ ] I have performed a self-review of my code, including formatting and typos
- [ ] I have [cleaned the commit history](https://www.annashipman.co.uk/jfdi/good-pull-requests.html)
- [ ] I have added the `Devops` label
- [ ] I have attached the pull request to the trello card

### CODE REVIEW

**Comment by RMcVelia** [comment]
[Create github workflow templates](https://trello.com/c/pxhaHtuQ/2088-create-github-workflow-templates)

File: build-docker-image/action.yml [hunk 1]
```diff
@@ -181,3 +183,15 @@ runs:
     - name: Push ${{ env.DOCKER_REPOSITORY }}:${{ env.IMAGE_TAG }} image to registry
       shell: bash
       run: docker image push --all-tags ${{ env.DOCKER_REPOSITORY }}
+
+    - name: Notify slack channel on job failure
+      uses: rtCamp/action-slack-notify@master
+      if: ${{ failure() && github.ref == 'refs/heads/main' }}
```
Comment (saliceti) [inline build-docker-image/action.yml:189]: inputs.slack-webhook is not required. Add a condition not to run when it's not provided?

File: deploy-to-aks/action.yml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    - uses: azure/login@v2
+      if: ${{ ( inputs.db-seed == 'true' && inputs.environment == 'review' ) }}
+      with:
+        creds: ${{ inputs.azure-credentials }}
+
+    - uses: google-github-actions/auth@v2
+      if: ${{ ( env.PROJECT_ID != '' && env.WIP != '' ) }}
+      with:
+        project_id: ${{ env.PROJECT_ID }}
```
Comment (saliceti) [inline deploy-to-aks/action.yml:73]: project_id may be optional actually: https://github.com/google-github-actions/auth?tab=readme-ov-file#inputs-miscellaneous
we can try without it

File: deploy-to-aks/action.yml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  using: composite
+  steps:
+    - name: Set Environment variables
+      id: set_env_var
+      shell: bash
+      run: |
+        terraform_version=$(awk '/{/{f=/^terraform/;next}f' ${{ inputs.terraform-base }}/${{ inputs.terraform-provider }} | grep -o [0-9\.]*)
+        echo "TERRAFORM_VERSION=$terraform_version" >> $GITHUB_ENV
+        echo "GCP_PROJECT_ID=$(make -s printvar-GCP_PROJECT_ID || true)" >> $GITHUB_ENV
+        echo "GCP_WIP=$(make -s printvar-GCP_WIP || true)" >> $GITHUB_ENV
```
Comment (saliceti) [inline deploy-to-aks/action.yml:57]: This long string is only used for workflows. To simplify, could we try a reposiroty variable instead? https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#creating-configuration-variables-for-a-repository

File: deploy-to-aks/action.yml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      run: make ${{ inputs.environment }} ci terraform-apply
+      env:
+        IMAGE_TAG: ${{ inputs.sha }}
+        PR_NUMBER: ${{ inputs.pr-number }}
+
+    - name: Install kubectl
+      if:  inputs.pr_number != '' && inputs.seed-cmd != ''
+      uses: DFE-Digital/github-actions/set-kubectl@master
+
+    - name: Seed review app
```
[the remark this answers is not in the document]
Comment (RMcVelia) [reply deploy-to-aks/action.yml:91]: done

File: deploy-to-aks/action.yml [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    required: false
+  terraform-url-output:
+    description: terraform output that contains the app urls
+    required: false
+    default: 'external_urls'
+  terraform-base:
+    description: path to the terraform files
+    required: false
+    default: 'terraform/application'
+  terraform-provider:
```
Comment (saliceti) [inline deploy-to-aks/action.yml:42]: Seems easy to make the same for all repos. Can we hardcode to terraform.tf and fix repos going forward?

File: deploy-to-aks/action.yml (hunk 3 above)
Comment (saliceti) [reply deploy-to-aks/action.yml:57]: Or pass it as input as discussed before

File: deploy-to-aks/action.yml [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    default: false
+  smoke-test:
+    description: Enable smoke test after deployment
+    type: boolean
+    required: false
+    default: false
+  healthcheck:
+    description: Health check path
+    required: false
+  terraform-url-output:
```
Comment (saliceti) [inline deploy-to-aks/action.yml:34]: As discussed we can make the output always called external_urls, so can we hardcode it?

File: deploy-to-aks/action.yml [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        echo "TERRAFORM_VERSION=$terraform_version" >> $GITHUB_ENV
+
+    - name: Use Terraform ${{ env.TERRAFORM_VERSION }}
+      uses: hashicorp/setup-terraform@v3
+      with:
+        terraform_version: ${{ env.TERRAFORM_VERSION }}
+        terraform_wrapper: false
+
+    - uses: azure/login@v2
+      if: ${{ ( inputs.db-seed == 'true' && inputs.environment == 'review' ) }}
```
Comment (saliceti) [inline deploy-to-aks/action.yml:63]: Can we make it the responsibility of the caller so we don't hardcode the environment name and allow more flexibility
e.g. `seed-review-app: ${{ inputs.environment == review }}`

Comment (RMcVelia) [reply deploy-to-aks/action.yml:63]: If you mean to remove the ${{ inputs.environment == review }} I would just be concerned someone could accidentally seed a non-review app, even production.

File: deploy-to-aks/action.yml [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      run: make ci ${{ inputs.environment }} db-seed
+      env:
+        PR_NUMBER: ${{ inputs.pr-number }}
+
+    - name: set DEPLOY_URL output
+      shell: bash
+      id: set_output
+      run: |
+        first_url=$(terraform -chdir=${{ inputs.terraform-base }} output -json external_urls | jq -r '.[0]')
+        echo "DEPLOY_URL=${first_url}" >> $GITHUB_OUTPUT
```
Comment (saliceti) [inline deploy-to-aks/action.yml:99]: Can we rename to ENVIRONMENT_URL for consistency with the output?

File: deploy-to-aks/action.yml (hunk 7 above)
Comment (RMcVelia) [reply deploy-to-aks/action.yml]: removed

File: deploy-to-aks/action.yml [hunk 9]
```diff
@@ -0,0 +1,147 @@
+name: Deploy to AKS environment
+description: Deploy a docker image to an AKS environment
+
+inputs:
```
Comment (saliceti) [inline deploy-to-aks/action.yml:4]: Can we add the missing descriptions?

File: deploy-to-aks/action.yml (hunk 8 above)
Comment (RMcVelia) [reply deploy-to-aks/action.yml]: done

File: deploy-to-aks/action.yml (hunk 9 above)
Comment (RMcVelia) [reply deploy-to-aks/action.yml:4]: added missing descriptions

File: deploy-to-aks/action.yml [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+  azure-credentials:
+    description: A JSON string containing service principle credentials.
```
Comment (saliceti) [inline deploy-to-aks/action.yml:6]: "principal"

File: deploy-to-aks/README.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+Optionally after deployment
+- run a seed command for review apps
+  - if enabled, it will run 'make review seed-review-app' so this must exist in the Makefile
+- run a smoktest after deployment
+
+If using Google Cloud then GCP_PROJECT_ID abd GCP_WIP variables must be set in the service Makefile.
+
+## Inputs
+- `azure-credentials`: A JSON string containing service principle credentials (Required)
+- `gcp-wip`: The full identifier of the GCP Workload Identity Provider (Optional)
```
Comment (saliceti) [inline deploy-to-aks/README.md:13]: Link to https://github.com/DFE-Digital/terraform-modules/blob/main/aks/dfe_analytics/README.md#authentication---github-actions

File: deploy-to-aks/action.yml [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    description: PR number for the review app
+    required: false
+  sha:
+    description: Commit sha to be deployed
+    required: true
+  slack-webhook:
+    description: Name of the slack webhook
+    required: false
+  db-seed:
+    description: Run seed command after a deployment
```
Comment (saliceti) [inline deploy-to-aks/action.yml:27]: Maybe repeat it's normally only for review apps

File: deploy-to-aks/README.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+- `sha`: commit sha of the docker image to be deployed (Required)
+- `slack-webhook` : A slack webhook to send a slack message to the service tech channel on deploy failure (Optional)
+- `smoke-test` : Run an application smoke test after deployment (default: false)
+- `healthcheck` : Health check path (Optional)
+- `terraform-base` : Name of the base terraform path (default: 'terraform/application')
+
+## Example
+
+```yaml
+- name: Checkout
```
Comment (saliceti) [inline deploy-to-aks/README.md:27]: Could we include checkout in the action?

File: deploy-to-aks/README.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+## Inputs
+- `azure-credentials`: A JSON string containing service principle credentials (Required)
+- `gcp-wip`: The full identifier of the GCP Workload Identity Provider (Optional)
+- `environment`: Name of the environment to deploy (Required)
+- `github-token`: Default Github token retrieved via secrets. GITHUB_TOKEN or PAT with permission to the repository (Required)
+- `pr-number`: Pull Request Number if deploying a review app (Optional)
+- `db-seed`: Run seed command after a deployment. Should only be used for review apps (default: false)
+- `sha`: commit sha of the docker image to be deployed (Required)
+- `slack-webhook` : A slack webhook to send a slack message to the service tech channel on deploy failure (Optional)
```
Comment (saliceti) [inline deploy-to-aks/README.md:19]: Add link to tech guidance?

File: deploy-to-aks/README.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+If using Google Cloud then GCP_PROJECT_ID abd GCP_WIP variables must be set in the service Makefile.
+
+## Inputs
+- `azure-credentials`: A JSON string containing service principle credentials (Required)
+- `gcp-wip`: The full identifier of the GCP Workload Identity Provider (Optional)
+- `environment`: Name of the environment to deploy (Required)
+- `github-token`: Default Github token retrieved via secrets. GITHUB_TOKEN or PAT with permission to the repository (Required)
+- `pr-number`: Pull Request Number if deploying a review app (Optional)
+- `db-seed`: Run seed command after a deployment. Should only be used for review apps (default: false)
+- `sha`: commit sha of the docker image to be deployed (Required)
```
Comment (saliceti) [inline deploy-to-aks/README.md:18]: Maybe more explicit:
Commit sha corresponding to the docker image tag to be deployed (Required)

File: deploy-to-aks/README.md [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+- `slack-webhook` : A slack webhook to send a slack message to the service tech channel on deploy failure (Optional)
+- `smoke-test` : Run an application smoke test after deployment (default: false)
+- `healthcheck` : Health check path (Optional)
```
Comment (saliceti) [inline deploy-to-aks/README.md:21]: Maybe mention it's without first slash (give an example)

File: deploy-to-aks/README.md [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+  uses: DFE-Digital/github-actions/deploy-to-aks@master
+  with:
+    azure-credentials:  ${{ secrets.AZURE_CREDENTIALS }}
+    environment: review
+    github-token: ${{ secrets.GITHUB_TOKEN }}
+    pr-number: ${{ github.event.pull_request.number }}
+    sha: ${{ needs.build.outputs.docker-image-tag }}
+    healthcheck: 'healthcheck/all'
+    db-seed: true
+```
```
Comment (saliceti) [inline deploy-to-aks/README.md:41]: Can we add gcp-wif and smoke-test since this will be the normal case?

File: deploy-to-aks/action.yml (hunk 10 above)
Comment (RMcVelia) [reply deploy-to-aks/action.yml]: fixed

File: deploy-to-aks/README.md (hunk 11 above)
Comment (RMcVelia) [reply deploy-to-aks/README.md]: added link

File: deploy-to-aks/action.yml [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+  azure-credentials:
+    description: A JSON string containing service principle credentials.
+    required: true
+  gcp-wip:
+    description: The full identifier of the GCP Workload Identity Provider
+    required: false
+  environment:
+    description: Environment to deploy to
+    required: true
+  github-token:
```
[the remark this answers is not in the document]
Comment (RMcVelia) [reply deploy-to-aks/action.yml]: removed

File: deploy-to-aks/action.yml (hunk 12 above)
Comment (RMcVelia) [reply deploy-to-aks/action.yml]: added note

File: deploy-to-aks/README.md (hunk 13 above)
Comment (RMcVelia) [reply deploy-to-aks/README.md:26]: moved to action

File: deploy-to-aks/README.md (hunk 14 above)
Comment (RMcVelia) [reply deploy-to-aks/README.md]: added link

File: deploy-to-aks/README.md (hunk 15 above)
Comment (RMcVelia) [reply deploy-to-aks/README.md]: done

File: deploy-to-aks/README.md (hunk 17 above)
Comment (RMcVelia) [reply deploy-to-aks/README.md:39]: done

File: deploy-to-aks/README.md [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+## Example
+
+```yaml
+- name: Deploy App to Review
+  id: deploy_review
+  uses: DFE-Digital/github-actions/deploy-to-aks@master
+  with:
+    azure-credentials:  ${{ secrets.AZURE_CREDENTIALS }}
+    environment: review
+    github-token: ${{ secrets.GITHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (RMcVelia) [reply deploy-to-aks/README.md]: removed

**Review by saliceti — APPROVED** [review]

**Pull request merged**